Add `./manage.py process_emails_from_sqs` to continously process messages from SQS

jwhitlock opened the pull request:
This PR addresses Jira task MPP 1802, "Refactor email processing". It adds a management tasks that polls a queue (provided by [AWS Simple Queue Service](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/welcome.html), or SQS) and processes the messages. The queue is populated by the Simple Notification Service (SNS), and is an alternate processing path to the `POST`s to `/emails/sns_inbound`. Our hope is that processing emails this way will keep the website responsive during emails backlogs, such as ones around production pushes, and be a better fit for larger emails and more emails processing.

This adds two management commands, ``process_emails_from_sqs`` and ``check_health``

``./manage.py process_emails_from_sqs`` is heavily based on ``./manage.py process_delayed_emails_from_sqs``, but has been refactored for testability, monitoring, logging, and to run without stopping. The messages themselves appear identical to those placed on the dead-letter queue (DLQ), so this command could be used in place of ``process_delayed_emails_from_sqs``.

The command can periodically write JSON to a file, that includes a timestamp. This can be checked with ``./manage.py check_health [path_to_file]``, which returns a non-zero exit code if the file is too old. This can be used as a [liveness probe](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/) in Kubernetes, to restart the process if it gets stuck.

This includes new settings:

* ``AWS_SQS_EMAIL_QUEUE_URL`` - The URL of the SQS queue with the email messages
* ``AWS_SQS_EMAIL_DLQ_URL`` The "[dead-letter queue](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-dead-letter-queues.html)" (DLQ) for the SQS queue, when a message is retrieved but not deleted too many times. This is similar to ``AWS_SQS_QUEUE_URL``, the DLQ for the SNS ``POST`` process, and the same queue can be used.

This includes new gauges:

* ``email_queue_count`` - the approximate number of messages in the queue
* ``email_queue_delayed`` - the approximate number of delayed messages in the queue, newly added but not available to receive
* ``email_queue_count_not_visible`` - the approximate number of non-visible messages in the queue, recently sent to a receiver

These are tagged with `queue:name`, where `name` is taken from the last component of the SQS URL. This will allow us to use the same metric for the primary and the DLQ queues. 

This includes two requirements changes:
* Add `codetiming`, which is the package of the tutorial [Python Timer Functions: Three Ways to Monitor Your Code](https://realpython.com/python-timer/)
* Update `markus` to 4.0.0, to get the `assert_gauge` method and support Python 3.9 and 3.10.

This includes new documentation for setting up the SQS polling in the local development environment.

This PR includes tests for the new management commands, but mocks out any processing (`_sns_inbound_logic`) which is tested elsewhere.

How to test:

* Follow the existing end-to-end local development documentation to get email delivery working in your development environment. It may be necessary to get S3 storage of emails working as well (same docs).
* Follow the new steps in this PR to convert to SES → SNS → SQS, which includes updating settings and running ``./manage.py process_emails_from_sqs``

jwhitlock commented:
After writing the PR message, I realized I should tag the new gauge metrics, and added a commit to do that. I also added a section about the requirements changes.

This is a large PR! Maybe too large! If I need to break it up, I could potentially refactor ``process_delayed_emails_from_sqs`` over several PRs, to get to the same functionality. Other ideas are welcome.

groovecoder commented on emails/tests/mgmt_process_emails_from_sqs_tests.py:
```diff
@@ -0,0 +1,301 @@
+
+def test_main_loop_one_message():
+    """The main loop can process a message."""
+    msg = fake_sqs_message(json.dumps(EMAIL_SNS_BODIES["s3_stored"]))
```
praise: yay! all that work to create SNS message body fixtures paid off!!

groovecoder commented on emails/tests/mgmt_process_emails_from_sqs_tests.py:
```diff
@@ -0,0 +1,301 @@
+
+def test_main_loop_ses_temp_failure_twice(mock_sns_inbound_logic):
+    """A temporary error followed by a second error is not retried."""
+    temp_error = ClientError(
+        {
+            "Error": {
+                "Message": "Maximum sending rate exceeded.",
+                "Code": "ThrottlingException",
+            }
+        },
+        "",
+    )
```
question (non-blocking): in `pytest`, could this be made into its own fixture that is requested in each of these 2 tests that needs it? to eliminate the code duplication?

jwhitlock replied:
I created a function `make_client_error` to help construct the three `ClientErrors` tested, and switched the second one to test `AccountSendingPausedException`, which I'm guessing is the other temporary exception that has the phrase 'pause'

And while I'm writing this, I see one more ``ClientError`` that could use ``make_client_error``...

groovecoder commented on emails/tests/mgmt_process_emails_from_sqs_tests.py:
```diff
@@ -0,0 +1,301 @@
+    AWS_REGION="us-east-2",
+    AWS_SQS_EMAIL_QUEUE_URL="https://sqs.us-east-2.amazonaws.example.com/111222333/queue-name",
+)
+def test_command_golden_path(mock_boto3_queue_constructor):
+    """The command runs successfully until completion."""
+    mock_boto3_queue_constructor.return_value = fake_queue()
+    call_command("process_emails_from_sqs", "--max-seconds=4")
+    mock_boto3_queue_constructor.assert_called_once_with(
+        "https://sqs.us-east-2.amazonaws.example.com/111222333/queue-name"
+    )
+    assert mock_boto3_queue_constructor._mock_region == "us-east-2"
```
suggestion (non-blocking): How about a `super_golden_path` with lots of messages that all succeed, and a `dumpster_fire_path` with lots of messages that all fail for different reasons?

jwhitlock replied:
In general, I don't like "a list of a lot of things" tests. It is unclear what code each individual message is testing, so you get code coverage without a clear example of how to trigger the code path. I find them to be harder to refactor, and often get uglier as time goes by. When they break, I often have to manually break up the list and run several times to determine which one broke, and how to fix them.

When a test really calls for several similar but slightly different inputs, I prefer `pytest.mark.parametrize`, so that each input is in a different test run. In many cases, this results in `if`/`then` clauses in the tests, which are another kind of test smell that I eliminated in the `check_health` tests.

The purpose of this test and the next is to get coverage of the argument parsing and the `handle` code. Most of the tests bypass these by directly calling the `main_loop`. Email processing is completely mocked out, and the tests for that are elsewhere, or need to be written.

Staring at this test, I feel that the test is poorly named. It runs for 4 (fake) seconds and processes no messages at all. I was expecting more cases when I named it, but it turned out there was just two branches to test. A rename is in order, something like `test_command_setup_success`

groovecoder reviewed (commented):
Whew, looking good!

Something I'm still not super clear about from the code or comments is the "cycle" vs. "loop" concepts? It seems like the "cycle" is the cycle of long polling, and the "loop" is the loop over queue messages?

jwhitlock commented on .env-dist:
```diff
@@ -17,6 +17,8 @@ AWS_SECRET_ACCESS_KEY=""
 AWS_SNS_TOPIC="arn:aws:sns:us-east-1:1234567890:ses-inbound"
 AWS_SES_CONFIGSET="dev_fxprivaterelay_env_dist"
 AWS_SQS_QUEUE_URL=""
+AWS_SQS_EMAIL_QUEUE_URL=""
+AWS_SQS_EMAIL_DLQ_URL=""
```
I think there will be a transition period with all three of them, but I think `AWS_SQS_QUEUE_URL` will go away in favor of these two.

jwhitlock commented on emails/management/commands/check_health.py:
```diff
@@ -0,0 +1,85 @@
+from datetime import datetime, timezone
+import json
+import logging
+import sys
```
Thanks! Is there a linter you're using?

jwhitlock commented on emails/management/commands/check_health.py:
```diff
@@ -0,0 +1,85 @@
+
+    def add_arguments(self, parser):
+        """Add command-line arguments (called by BaseCommand)"""
+        parser.add_argument(
+            "healthcheck_path",
+            help="Path to healthcheck JSON file",
+        )
```
The default type is a ``str``, and I usually omit it, but I could add for clarity. I first tried [FileType](https://docs.python.org/3/library/argparse.html#argparse.FileType), but I forget why I rejected that - I'll have to try it again to refresh my memory.

The default is to not write a healthcheck file, but something like `/tmp/relay_healthcheck.json` seems like a safe default if we'd like one.

jwhitlock replied:
I switched to `FileType`. I think an omitted default still makes sense to disable the feature. I haven't even gotten SRE feedback on the healthcheck yet, all the code may go away in the next PR...

jwhitlock commented on emails/tests/mgmt_check_health_tests.py:
```diff
@@ -0,0 +1,71 @@
+    with patch(
+        "emails.management.commands.check_health.logger", spec_set=("info", "error")
+    ) as mock_logger:
+        yield mock_logger
```
There is some implicit teardown logic. After the test completes, execution returns to the fixture. It then leaves the context manager, which undoes the patch to ``logger``.

I'm not sure where I picked up this technique. I've seen others sharing it, like https://stackoverflow.com/a/50048692/10612.

There's some alternatives. pytest provides [monkeypatch](https://docs.pytest.org/en/6.2.x/monkeypatch.html), which seems to me to be more awkward than ``unittest.mock.patch``. Maybe it was written before [mock](https://mock.readthedocs.io/en/latest/) was added to the standard library in 3.3

For this case, there is a [caplog fixture](https://docs.pytest.org/en/6.2.x/logging.html#caplog-fixture) that is probably better than rolling my own!

jwhitlock commented on emails/tests/mgmt_check_health_tests.py:
```diff
@@ -0,0 +1,71 @@
+
+
+@pytest.mark.parametrize("verbosity", (1, 2))
+def test_check_health_passed(tmp_path, mock_logger, verbosity):
```
`tmp_path` is one of the built-in pytest fixtures. The full list is in the fixtures doc: https://docs.pytest.org/en/6.2.x/fixture.html

groovecoder replied:
Whoa. So pytest checks thru all the parameter names to see if any match a built-in fixture? Does that also mean that parameter order doesn't matter in pytest?

jwhitlock commented on emails/tests/mgmt_check_health_tests.py:
```diff
@@ -0,0 +1,71 @@
+        json.dump(data, f)
+
+    call_command("check_health", str(path), f"--verbosity={verbosity}")
+    if verbosity == 1:
+        mock_logger.info.assert_not_called()
+    else:
```
I agree, I'm also suspicious of test code with if/else branches. I'm going to break this up from 2 to 4 tests, which will get rid of the branches, and switch to [caplog](https://docs.pytest.org/en/6.2.x/logging.html#caplog-fixture).

jwhitlock commented on emails/management/commands/process_emails_from_sqs.py:
```diff
@@ -0,0 +1,447 @@
+                with Timer(logger=None) as attribute_timer:
+                    self.queue.load()  # Refresh attributes
+                self.write_healthcheck()
+                incr_if_enabled("message_from_sqs_error", 1)
```
Good catch! This was copypasta from `process_message`, I'll remove.

jwhitlock commented on emails/management/commands/process_emails_from_sqs.py:
```diff
@@ -0,0 +1,447 @@
+        * success: True if message was processed successfully
+        * error: The processing error, omitted on success
+        * message_body_quoted: Set if the message was non-JSON, omitted for valid JSON
+        * json_body: Set to the JSON on SNS verification failures, or omitted on verified
```
Yep, makes sense to omit it until we decide we need to debug SNS verification failures

jwhitlock commented on emails/management/commands/process_emails_from_sqs.py:
```diff
@@ -0,0 +1,447 @@
+                    "email_queue_count_not_visible",
+                    self.queue.attributes["ApproximateNumberOfMessagesNotVisible"],
+                    tags=[generate_tag("queue", self.queue_name)],
+                )
```
I moved this code into `refresh_and_emit_queue_count_metrics`, as well as some other changes to the function to hopefully clarify things

jwhitlock commented on emails/management/commands/process_emails_from_sqs.py:
```diff
@@ -0,0 +1,447 @@
+                else:
+                    failed_count += 1
+                    if self.delete_failed_messages:
+                        message.delete()
```
I moved the deletion code out, which removes this branch. It now looks like:

```
                if not message_data["success"]:
                    failed_count += 1
                if message_data["success"] or self.delete_failed_messages:
                    message.delete()
```

jwhitlock commented:
Several changes from the previous review round:

In `docs/end-to-end-local-dev.md`:
* Documentation fix, "was are switching" to "we are switching"

In `emails/management/commands/check_health.py`:

* Switch ``healthcheck_path`` parameter from string to [argparse.FileType](https://docs.python.org/3/library/argparse.html#filetype-objects). There is no change for callers, but a few code changes to expect an open file instead of a string.

In `emails/management/commands/process_emails_from_sqs.py`:

* Switch ``--healthcheck_path`` from string to ``argparse.FileType`` as well
* Clarify that ``--sqs-url`` gets default from ``settings.AWS_SQS_EMAIL_QUEUE_URL``
* Rename method ``main_loop()`` to ``process_queue()``
* Move queue backlog metrics to method ``refresh_and_emit_queue_count_metrics``
* Split method ``cycle_loop` into ``poll_queue_for_messages()`` and ``process_message_batch``
* Remove ``json_body`` from logs when SNS verification fails
* Remove bad metric increment

In `emails/tests/mgmt_check_health_tests.py`:
* Migrate from custom fixture ``mock_logger`` to pytest fixture ``caplop``
* Move common healthcheck creation code to ``write_healthcheck``
* Split 2 tests with two variants into 4 simple tests

In `emails/tests/mgmt_process_emails_from_sqs_tests.py`:
* Removed unused import
* Add helper function `make_client_error` to help make `botocore.exceptions.ClientError` instances
* Rename tests and adjust docstrings for rename of `main_loop` to `process_queue`
* Add assertions that mocked methods are called when processing a message
* Rename `test_command_golden_path` to `test_command_successful_setup`

Open discussions:
* Should the log include items when they are zero (for example, `"pause_count": 0` when message processing did *not* pause for a retryable exception)?
* Should timers be float seconds (`sqs_poll_s=5.001`) or integer milliseconds (`sqs_poll_ms=5001`)?
* Should I add typing and mypy now or later?

There's some other (non-blocking) discussions, let me know if you want to continue them.

groovecoder commented on emails/management/commands/process_emails_from_sqs.py:
```diff
@@ -0,0 +1,447 @@
+        * pause_s: How long pauses took, omitted if no pauses
+        * failed_count: How many messages failed to process, omitted if 0
+
+        Times are in seconds, with millisecond precision
```
Okay cool. `_s=S.msms` is fine.

groovecoder reviewed (changes requested):
Looking good! I got it working with my local Relay app by following the new steps in the doc. Woohoo!

Just a couple more things:

suggestion: Let's add the S3 and SQS steps as optional in the "Overview" section?

```
6. (Optional) Convert to store in S3
7. (Optional) Convert to back-end processing
```

jwhitlock commented on docs/end-to-end-local-dev.md:
```diff
@@ -345,3 +345,115 @@ Same as before:
+    Sending MTA->>yourdomain.com: Fetch DNS MX Record
+    yourdomain.com->>Sending MTA: inbound-smtp.us-west-2.amazonaws.com.
+    Sending MTA->>AWS: SMTP
+    AWS<<-Local app: SQS fetch
+    Local app->>AWS: POST /SendRawEmail
```
ha mermaid can't read a arrow that goes left

jwhitlock commented:
Force push to pick up requirements merges on main, and then:

* Mark sections to setup S3 and SQS Push as optional
* Fix SQS push sequence diagram
